### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by tarasavage**

### CODE REVIEW

File: app/fight/knight.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            snd_knight: Knight
+    ) -> None:
+        if fst_knight.protection < snd_knight.power:
+            fst_knight.hp -= snd_knight.power - fst_knight.protection
+
+        if snd_knight.protection < fst_knight.power:
+            snd_knight.hp -= fst_knight.power - snd_knight.protection
+
+        if fst_knight.hp < 0:
+            fst_knight.hp = 0
```
Comment (ArsenPidhoretskyi) [inline app/fight/knight.py:62]: You can move this to method

File: app/fight/knight.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    def apply_weapon(self, weapon: Weapon) -> None:
+        self.power += weapon.power
+
+    @staticmethod
+    def fight(
+            fst_knight: Knight,
+            snd_knight: Knight
+    ) -> None:
+        if fst_knight.protection < snd_knight.power:
+            fst_knight.hp -= snd_knight.power - fst_knight.protection
```
Comment (ArsenPidhoretskyi) [inline app/fight/knight.py:56]: Also this

File: app/main.py [hunk 3]
```diff
@@ -1,214 +1,53 @@
...
+def battle(knights_config: dict) -> dict:
+    knights = {}
+
+    for knight_data in knights_config.values():
+        name = knight_data.get("name")
+        hp = knight_data.get("hp")
+        power = knight_data.get("power")
+        all_armour = [
+            Armour(armour["part"], armour["protection"])
+            for armour in knight_data.get("armour")
```
Comment (ArsenPidhoretskyi) [inline app/main.py:17]: .get can return None
Add default argument as list

File: app/main.py [hunk 4]
```diff
@@ -1,214 +1,53 @@
...
+        ]
+        weapon = knight_data.get("weapon")
+        potion = knight_data.get("potion")
+
+        params = [
+            name,
+            hp,
+            power,
+            all_armour,
+            Weapon(weapon["name"], weapon["power"])
```
Comment (ArsenPidhoretskyi) [inline app/main.py:27]: weapon can be None

File: app/fight/knight.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    def apply_potion(self, potion: Potion) -> None:
+        if "power" in potion.effect:
+            self.power += potion.effect["power"]
+
+        if "hp" in potion.effect:
+            self.hp += potion.effect["hp"]
+
+        if "protection" in potion.effect:
+            self.protection += potion.effect["protection"]
```
Comment (ArsenPidhoretskyi) [inline app/fight/knight.py:40]: Use loop for this and use setattr and getattr

File: app/fight/knight.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    def apply_armour(self, armour: list[Armour]) -> None:
+        self.protection = sum(
+            [armour_element.protection for armour_element in armour]
```
Comment (ArsenPidhoretskyi) [inline app/fight/knight.py:44]:
```suggestion
            armour_element.protection for armour_element in armour
```
No need list here

File: app/fight/knight.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        )
+
+    def apply_weapon(self, weapon: Weapon) -> None:
+        self.power += weapon.power
+
+    @staticmethod
+    def fight(
+            fst_knight: Knight,
```
Comment (ArsenPidhoretskyi) [inline app/fight/knight.py:52]: fst_knight is not so clear name
also with snd_knight

File: app/main.py [hunk 8]
```diff
@@ -1,214 +1,53 @@
...
+        ]
+
+        if potion is not None:
+            params.append(Potion(potion["name"], potion["effect"]))
+
+        new_knight = Knight(*params)
+        knights[name] = new_knight
+
+    lancelot = knights.get("Lancelot")
```
Comment (ArsenPidhoretskyi) [inline app/main.py:36]: .get can return None

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 3 above)
Comment (tarasavage) [reply app/main.py:17]: What should I do? I don't understand

File: app/main.py [hunk 9]
```diff
@@ -1,214 +1,55 @@
...
+def battle(knights_config: dict) -> dict:
+    knights = {}
+
+    for knight_data in knights_config.values():
+        name = knight_data.get("name")
+        hp = knight_data.get("hp")
+        power = knight_data.get("power")
+        all_armour = [
+            Armour(armour["part"], armour["protection"])
+            for armour in knight_data.get("armour")
```
Comment (Vasyl-Smutok) [inline app/main.py:17]:
```suggestion
            for armour in knight_data.get("armour", [])
```

File: app/main.py [hunk 10]
```diff
@@ -1,214 +1,55 @@
...
+        ]
+        weapon = knight_data.get("weapon")
+        potion = knight_data.get("potion")
+
+        params = [
+            name,
+            hp,
+            power,
+            all_armour,
+        ]
```
Comment (Vasyl-Smutok) [inline app/main.py:27]:
```suggestion
        params = [name, hp, power, all_armour,]
```

**Review by Vasyl-Smutok — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 3 above)
Comment (Vasyl-Smutok) [reply app/main.py:17]: image

**Pull request closed without merging**